tools/test/LcovMerger/java/com/google/devtools/lcovmerger/Coverage.java

```diff
@@ -45,6 +46,25 @@ static Coverage merge(Coverage c1, Coverage c2) {
     return merged;
   }
 
+  static Coverage filterOutMatchingSources(
+      Coverage coverage, String[] regex) throws IllegalArgumentException {
+    if (coverage == null || regex == null) {
+      throw new IllegalArgumentException("Coverage and regex should not be null.");
+    }
+    if (regex.length == 0) {
+      return coverage;
+    }
+    Coverage filteredCoverage = new Coverage();
+    coverage.getAllSourceFiles().stream()
+            .filter(s -> !matchesAnyRegex(s.sourceFileName(), regex))
+            .forEach(s -> filteredCoverage.add(s));
+    return filteredCoverage;
+  }
+
+  private static boolean matchesAnyRegex(String input, String[] regex) {
+    return Arrays.stream(regex).anyMatch(r -> input.matches(r));
```

Comment:
I think constructing a stream is overkill here. Not sure about performance (though https://github.com/bazelbuild/bazel/commit/5f39475bb51f48182180e34e4310edb650bea462 hints at a performance hit), but it surely creates some garbage. A C-style for-loop would be easily readable here:

    for (int i = 0; i < regex.length; i++) {
      if (regex[i].matches(input)) {
        return true;
      }
    }
    return false;

WDYT?

Reply:
Do we have some data to show how C-style for-loops are better, or some way to compare performance between these two approaches? My previous code reviewers had push strongly for using streams and lambdas. We may need some guidelines on this topic otherwise we'll end up with inconsistent and maybe under-performing code.

Reply:
I don't have data, just these PRs: https://github.com/bazelbuild/bazel/pull/5737, https://github.com/bazelbuild/bazel/pull/5735.

Reply:
I have data now: https://github.com/laszlocsomor/projects/tree/master/profile/java-iterators-speed

50M-element String array iteration is faster with iterators than with C-style loop, and C-style loop is faster than streams.

Reply:
Thinking about this more... I don't know of high resolution clocks for Java so it only makes sense to measure many iteration cycles, but then the JVM will detect and optimize these hot spots in the code.

Speed is likely also dependent on the number of elements.

Bottom line is: unless you have tens of millions of items or more, it probably doesn't matter which iteration approach you use, they'll be about the same speed.

Measuring memory usage would be interesting though, but I don't want to block the commit on this.

Reply:
Wow, thanks for doing the performance measurements! I will use them in the feature as reference. 

I used iterator loops in the end since it was the faster solution and still readable.